onnxscript/optimizer/_inliner.py

```diff
@@ -0,0 +1,272 @@
+        if new_name is not None:
+            new_name = _make_unique_name(new_name, self._call_stack, self._inliner.used_node_names)
+
+        new_metadata = {**self._metadata_props, **copied.metadata_props}
+        # TODO: For now, node metadata overrides callnode metadata if there is a conflict.
+        # Do we need to preserve both?
+
+        new_node = ir.Node(
+            copied.domain,
+            copied.op_type,
+            new_inputs,
+            new_attributes,
+            overload=copied.overload,
+            num_outputs=len(copied.outputs),
+            graph=None,  #  TODO:
+            name=new_name,
+            doc_string=copied.doc_string,
+            metadata_props=new_metadata,
+        )
+        new_outputs = new_node.outputs
+        for i, output in enumerate(copied.outputs):
+            self._value_map[output] = new_outputs[i]
+            old_name = output.name if output.name is not None else f"output_{i}"
+            new_outputs[i].name = _make_unique_name(old_name, self._call_stack, self._inliner.used_value_names)
+
+        self._inliner.node_context[new_node] = self._call_stack
+
+        return new_node
+
+    def clone_graph(self, graph: ir.Graph) -> ir.Graph:
+        input_values = [self.clone_value(v) for v in graph.inputs]
+        nodes = [self.clone_node(node) for node in graph]
+        initializers = graph.initializers  # Initializers are not cloned, but shared.
+
+        return ir.Graph(
+            input_values,  # type: ignore
+            graph.outputs,
+            nodes=nodes,
+            initializers=list(initializers.values()),
+            doc_string=graph.doc_string,
+            opset_imports=graph.opset_imports,
+            name=graph.name,
+            metadata_props=graph.metadata_props,
+        )
+
+def _abbreviate(function_ids: Iterable[ir.OperatorIdentifier]) -> dict[ir.OperatorIdentifier, str]:
+    """Create a short unambiguous abbreviation for all function ids."""
+    def id_abbreviation(id: ir.OperatorIdentifier) -> str:
+        """Create a short unambiguous abbreviation for a function id."""
+        domain, name, overload = id
+        # Omit the domain, if it remains unambiguous after omitting it.
+        if any (x[0] != domain and x[1] == name and x[2] == overload for x in function_ids):
+            short_domain = domain + "_"
+        else:
+            short_domain = ""
+        if overload != "":
+            return short_domain + name + "_" + overload
+        return short_domain + name
+    return {id: id_abbreviation(id) for id in function_ids}
+
+class _Inliner:
+    def __init__(self, model: ir.Model) -> None:
+        self._functions = model.functions
+        self._function_id_abbreviations = _abbreviate(self._functions.keys())
+        self._opset_imports = model.opset_imports
+        self.used_value_names: set[str] = set()
+        self.used_node_names: set[str] = set()
+        self.node_context: dict[ir.Node, CallStack] = {}
```

Comment:
Can they be made private?

Reply:
They are used in the auxiliary class _CopyReplace above. I made this class (Inliner) itself private, so hopefully it is clear that this is meant to be internal to the file (but usable by other classes in the file). Not sure if there is a better convention for this.

Reply:
I see - I didn't realized they are used outside. Thanks!